pepper-apis/src/main/java/org/broadinstitute/ddp/util/RouteUtil.java

```diff
@@ -149,4 +159,18 @@ public static ActivityInstanceDto findAccessibleInstanceOrHalt(Response response
 
         return instanceDto;
     }
+
+    /**
+     * Figures out a preferred user language taking into account the language weights
+     * from Accept-Language header, information from the user profile and languages
+     * supported by the study.
+     */
+    public static Locale resolvePreferredUserLanguage(
+            Handle handle, String acceptLanguageHeader, Locale preferredLocale, String studyGuid
+    ) {
+        List<LanguageRange> acceptLanguages = StringUtils.isNotEmpty(acceptLanguageHeader)
+                ? LanguageRange.parse(acceptLanguageHeader) : new ArrayList<>();
+        Set<Locale> localesSupportedByStudy = handle.attach(StudyDao.class).getSupportedLocalesByGuid(studyGuid);
+        return I18nUtil.resolvePreferredLanguage(preferredLocale, acceptLanguages, localesSupportedByStudy);
```

Comment:
I don't see changes in `I18nUtil` in this PR. Currently, `I18nUtil` won't behave the way we want. It will prefer the `preferredLocale` over `acceptLanguages` locales, whereas we want `acceptLanguages` to have higher priority than `preferredLocale`.